Related data mouseover

ItIsJordan opened the pull request:
* Adds mouseover tooltip support to the related data feature.  Closes #691.
* Improve the linking for records/tables for records with multiple versions.

ItIsJordan committed:
Add new related data functions - wip

Adds new functions to the HEPSubmission and DataSubmission  class to query for other related data object. Also adds a function to generate a dictionary generator function for combining recid/title data for the tooltip feature.

ItIsJordan committed:
Use new related data query functions

Updates the code to use the new related data functions.

ItIsJordan committed:
Update record page related link text

Updates the related page link text to alter the record section to display record ID, with the title as a tooltip. Updates table DOI links to display the table name, with description as a tooltip.

ItIsJordan committed:
wip test update for related records

Updates the testing for related record linking, currently working out an issue with results collection for checking.

ItIsJordan committed:
Merge branch 'main' into related-data-mouseover

ItIsJordan committed:
Move title to anchor tag in related recid html

Moves the title to the anchor tag for consistency with the other related data elements.

ItIsJordan committed:
Fix error with function in DataSubmission

Fixes an error with the get_related_datasubmissions function in DataSubmission where the logic was incorrect. Also fixes the DataSubmission.get_table_data_list to use the correct relation get function.

ItIsJordan committed:
Add unit test update for new related functions (for tooltips)

Adds tests for the related and related_to_this functions for DataSubmission and HEPSubmission objects

ItIsJordan committed:
Add e2e tests for related data tooltip update

Adds e2e tests to check the new tooltip update works for related data for records and tables.

ItIsJordan committed:
Merge branch 'main' into related-data-mouseover

ItIsJordan committed:
Clean/fix record/table data dict functions

Removes an unnecessary query in the get_table_data_list and get_record_data_list functions. Cleans up the variables, and removes a redundant check.

ItIsJordan committed:
Cleanup and commenting

Adds comments to test functions, small cleaning changes to the new feature code.

ItIsJordan committed:
Merge branch 'main' into related-data-mouseover

ItIsJordan committed:
Merge branch 'main' into related-data-mouseover

ItIsJordan committed:
Merge branch 'main' into related-data-mouseover

ItIsJordan committed:
Related function refactoring

Moves functions for related data from the models.py file and common.py into /records/api.py. Updates test_records.py to use new functions

ItIsJordan committed:
Update related cleanup function

Updates the related cleanup function to use the get_latest_hepsubmission function to ensure it always gets the most recent version, as prior versions may/will not have related data.

ItIsJordan committed:
Merge branch 'main' into related-data-mouseover

ItIsJordan committed:
Update regex for url check in related testing

Updates a regular expression for checking url within a webpage to be more open in e2e/test_records.py:test_related_records

ItIsJordan committed:
Add partial test for version tooltip testing

Adds a partially completed test for version tooltip testing in the e2e testing. Also adds the relevant testing data.

ItIsJordan committed:
Add finished test for related version tooltips

Adds finished test to check functionality of record page related table link tooltips.

ItIsJordan committed:
Fix broken test in test_records

Fiix broken path in related record test in test_records

ItIsJordan committed:
Fix path in test record

Fixes path in test_records that breaks during full test run

ItIsJordan committed:
FIx docstring in new functions

Fixes docstring in newly added functions.

ItIsJordan committed:
Update commenting

Updates some comments in branch changes

GraemeWatt commented on hepdata/modules/records/api.py:
```diff
@@ -36,6 +35,7 @@
 from invenio_accounts.models import User
 from invenio_db import db
 from sqlalchemy import and_
+from sqlalchemy.orm import aliased
```
The import `aliased` doesn't seem to be used.

GraemeWatt commented on hepdata/modules/records/api.py:
```diff
@@ -1043,3 +1043,139 @@ def get_all_ids(index=None, id_field='recid', last_updated=None, latest_first=Fa
+            HEPSubmission.query
+            .filter(HEPSubmission.publication_recid == related.related_recid)
+            .filter(HEPSubmission.overall_status == 'finished').first()
+        )
```
What if the record `related.related_recid` has multiple finished versions?  We only want to link to the latest one.  You can do it by either replacing `.first()` with `.order_by(HEPSubmission.version.desc()).first()` or using:
```python
data_submission = get_latest_hepsubmission(publication_recid=related.related_recid, overall_status='finished')
```

GraemeWatt commented on hepdata/modules/records/api.py:
```diff
@@ -1043,3 +1043,139 @@ def get_all_ids(index=None, id_field='recid', last_updated=None, latest_first=Fa
+        .join(RelatedRecid, RelatedRecid.this_recid == HEPSubmission.publication_recid)
+        .filter(RelatedRecid.related_recid == submission.publication_recid)
+        .filter(HEPSubmission.overall_status == 'finished')  # Only finished submissions
+        .all()
```
Again, for the case that there are multiple finished records with the same `recid`, only use the latest version.

GraemeWatt commented on hepdata/modules/records/api.py:
```diff
@@ -1043,3 +1043,139 @@ def get_all_ids(index=None, id_field='recid', last_updated=None, latest_first=Fa
+              & (HEPSubmission.version == DataSubmission.version))
+        .filter(RelatedTable.related_doi == data_submission.doi)
+        .filter(HEPSubmission.overall_status == 'finished')
+        .all()
```
Check that the `HEPSubmission` object corresponding to the `table_doi` is the latest finished version in case there are more than one.  Links where the `table_doi` is from an earlier version of a record should not be displayed.

GraemeWatt commented on hepdata/modules/records/assets/js/hepdata_tables.js:
```diff
@@ -203,13 +205,21 @@ HEPDATA.table_renderer = {
+        var related_object = relatedDois[i];
+        // Create the the `li` tag, setting the text to the table name and DOI
+        // Adds a `title` tag containing the table's description
+        // Example: "Table 1 - (10.17182/hepdata.322.v1/t1)"
```
Update the example to just "Table 1" (you don't display the DOI now).

GraemeWatt reviewed (changes requested):
Thanks, this is great.  I left a few comments, mainly related to the subtleties of linking to records with multiple finished versions, which we discussed at our last meeting in my office.

ItIsJordan committed:
Update related data testing

Updates related data testing in test_records as expected rendering has changed.

ItIsJordan committed:
Fix records test pathing

Fixes directory paths in test_records.py related version testing.

ItIsJordan committed:
Fix commenting

Fixes comment for doi example in hepdata_tables.js

ItIsJordan committed:
Update api.py (remove unused import)

Removes unused import from api.py

ItIsJordan committed:
Update record api related data functions

Updates the related data functions in records/api.py to be request latest versions where possible

ItIsJordan committed:
Update test_records.py

Updates pathing in related data testing

ItIsJordan committed:
Update related function

Makes the set used in get_related_to_this_hepsubmissions much simpler than before

ItIsJordan committed:
Fix e2e related record test

Fixes the e2e related record test to dynamically generate the ids to ensure functionality when ran standalone and alongside the test suite

sjmf commented on hepdata/modules/records/api.py:
```diff
@@ -57,7 +56,8 @@
 from hepdata.modules.records.utils.yaml_utils import split_files
 from hepdata.modules.stats.views import increment, get_count
-from hepdata.modules.submission.models import RecordVersionCommitMessage, DataSubmission, HEPSubmission, DataReview
+from hepdata.modules.submission.models import RecordVersionCommitMessage, DataSubmission, HEPSubmission, DataReview, \
```
Perhaps given the number of imports here:
```
from hepdata.modules.submission.models import \
    RecordVersionCommitMessage,
    DataSubmission,
    HEPSubmission,
    DataReview,
    RelatedRecid,
    RelatedTable
```

sjmf replied:
(backslash may not be needed after `import` now I think about it)

sjmf commented on hepdata/modules/records/api.py:
```diff
@@ -262,8 +262,8 @@ def format_resource(resource, contents, content_url):
-    ctx['related_recids'] = [r.related_recid for r in hepsubmission.related_recids]
-    ctx['related_to_this_recids'] = [s.publication_recid for s in hepsubmission.get_related_hepsubmissions()]
+    ctx['related_recids'] = get_record_data_list(hepsubmission, "related")
+    ctx['related_to_this_recids'] = get_record_data_list(hepsubmission, "related_to_this")
```
This is good, more readable than before.

sjmf commented on hepdata/modules/records/api.py:
```diff
@@ -1043,3 +1042,140 @@ def get_all_ids(index=None, id_field='recid', last_updated=None, latest_first=Fa
+        .filter(RelatedRecid.related_recid == submission.publication_recid)
+        .filter(HEPSubmission.overall_status == 'finished')  # Only finished submissions
+        .with_entities(HEPSubmission.publication_recid)
+        .all())
```
Can we make the indentation more consistent? Here we have:
```related_sub_ids = (HEPSubmission.query```

But below on line 1100 it's:
```
submission = (
            DataSubmission.query.filter(DataSubmission.doi == related.related_doi)
                .join(HEPSubmission, HEPSubmission.publication_recid == DataSubmission.publication_recid)
```
And then below that on line 1120 the indentation of the functions are all in line with the first statement, where the others examples above were indented:
```
related_submissions = (
        DataSubmission.query.join(RelatedTable, RelatedTable.table_doi == DataSubmission.doi)
        .join(HEPSubmission,(HEPSubmission.publication_recid == DataSubmission.publication_recid))
        .group_by(DataSubmission.id)
```

I would say just try and think about how to keep these consistent. The code style I would use would be: Indent subsquent lines, indent anything after `.query.`, intent after bracket. So:

```
related_sub_ids = (
        HEPSubmission.query
                .join(RelatedRecid, RelatedRecid.this_recid == HEPSubmission.publication_recid)
                .filter(RelatedRecid.related_recid == submission.publication_recid)
                .filter(HEPSubmission.overall_status == 'finished')  # Only finished submissions
                .with_entities(HEPSubmission.publication_recid)
                .all()
)
```

As long as they're all the same that's probably fine but I reckon the above is more readable.

sjmf commented on tests/submission_test.py:
```diff
@@ -33,7 +33,9 @@
-from hepdata.modules.records.api import format_submission, process_saved_file, create_new_version
+from hepdata.modules.records.api import format_submission, process_saved_file, create_new_version, get_record_data_list, \
+    get_related_hepsubmissions, get_related_datasubmissions, get_related_to_this_datasubmissions, get_table_data_list, \
+    get_related_to_this_hepsubmissions
```
As per comment on `api.py`, this is a bit of a run-on import statement which may be more readable spread over more lines.

sjmf commented on tests/submission_test.py:
```diff
@@ -327,13 +329,21 @@ def test_related_records(app, admin_idx):
         ]
         # Dummy record data
-        record = {'title': 'HEPData Testing',
+        record = {'title': None,
```
What was the rationale for replacing the title with `None`?

GraemeWatt replied:
Jordan's comment: "The title will be set later based on `test_data` values."

sjmf reviewed (changes requested):
A good PR. Have left a couple of comments to address in there, possibly with another commit.

ItIsJordan committed:
Related record test comment update

Updates commenting in test_related_records in submission_test.py to clarify a change to the dummy data

ItIsJordan committed:
Fix indentation

Fixes indentation for consistency in the related data functions in records/api.py.

ItIsJordan committed:
Clean import statements

Cleans up some lengthy import statements in submission_test.py, and records/api.py

GraemeWatt committed:
Merge branch 'main' into related-data-mouseover
